### ISSUE TITLE
Compiler allows incorrect interface implementation

### ISSUE THREAD

**Issue by eugene-korobko** [microsoft/TypeScript#48085]
# Bug Report

### 🔎 Search Terms

'incorrect interface implementation'
'interace with enum param'

### 🕗 Version & Regression Information

I've tried it with playground and it looks like the bug exists in all the versions.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.9.7#code/JYOwLgpgTgZghgYwgAgJLIN4Chm+TAVxAAoBnALmVLClAHMBKSgNwHtgATAbiwF8ssEEAQC2yAMo16ANTgAbUpiwBIZgEZkAXmQBydToA0K5gCYtu04b4CEcuKUXjkwEQAc5EEULCL02PMiuBABGcsAI+ERklJK0IHSyCkzIbJxKAQEIrCCkrB4AdHKsdGQMPAH8-EA)

### 💻 Code

```ts
interface I {
    fun(s: string): void;
}

enum StringVals {
	v1 = 'v1',
	v2 = 'v2',
}

class S implements I {
    public fun(s: StringVals): void {
        console.log(s);
    }
}
```

This leads to very strange behavior. Imagine we have code

```ts
function factory(): S {
    return new S();
}
```

It works nice. If we try

```ts
factory().fun('x')
```
compiler will show an error.

now we rewrite the function

```ts
function factory(): I {
    return new S();
}
```

and this change makes the compiler interpret code

```ts
factory().fun('x')
```

as correct. I believe that changing resulting type of factory method must not lead to type check weakening.

### 🙁 Actual behavior

Compiler successfully compiles the code

### 🙂 Expected behavior

Compile should show an error about incorrect implementation fun method in S class

**Comment by eugene-korobko** [comment]
Oh, thank you forclarification.
It looks like I've been affected with quite rare case that has been sacrificed for more common and popular cases.

**Comment by eugene-korobko** [comment]
IMHO, the perfect solution would be allowing conversion 
```
readonly Dog[] -> readonly Animal[]
```

and prohibiting
```
Dog[] -> Animal[]
```
